searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp

```diff
@@ -213,57 +213,96 @@ class Loader {
 };
 }
 
-
+/**
+ * Will load and index documents in parallel. Note that indexing order is not guaranteed,
+ * but that is inline with the guarantees vespa already has.
+ */
 class DenseTensorAttribute::ThreadedLoader : public Loader {
 public:
     ThreadedLoader(DenseTensorAttribute & attr, vespalib::Executor & shared_executor)
         : _attr(attr),
           _shared_executor(shared_executor),
-          _complete_executor(1, 0x10000),
+          _queue(MAX_PENDING),
           _pending(0)
     {}
-    void load(uint32_t lid, vespalib::datastore::EntryRef ref) override {
+    void load(uint32_t lid, vespalib::datastore::EntryRef ref) override;
+    void wait_complete() override {
+        drainUntilPending(0);
+    }
+private:
+    using Entry = std::pair<uint32_t, std::unique_ptr<PrepareResult>>;
+    using Queue = vespalib::ArrayQueue<Entry>;
+
+    bool pop(Entry & entry) {
+        std::unique_lock guard(_mutex);
+        if (_queue.empty()) return false;
+        entry = std::move(_queue.front());
+        _queue.pop();
+        return true;
+    }
+    void drainQ() {
+        Queue queue(MAX_PENDING);
         {
             std::unique_lock guard(_mutex);
-            while (_pending > MAX_PENDING) {
-                _cond.wait(guard);
-            }
+            queue.swap(_queue);
+        }
+        while (!queue.empty()) {
+            auto item = std::move(queue.front());
+            queue.pop();
+            complete(item.first, std::move(item.second));
         }
-        ++_pending;
-        _shared_executor.execute(vespalib::makeLambdaTask([this, ref, lid]() {
-            auto prepared = _attr._index->prepare_add_document(lid, _attr._denseTensorStore.get_typed_cells(ref), _attr.getGenerationHandler().takeGuard());
-            _complete_executor.execute(vespalib::makeLambdaTask([this, lid, result=std::move(prepared)]() mutable {
-                _attr.setCommittedDocIdLimit(std::max(_attr.getCommittedDocIdLimit(), lid + 1));
-                _attr._index->complete_add_document(lid, std::move(result));
-                {
-                    std::unique_lock guard(_mutex);
-                    --_pending;
-                    if ((_pending == MAX_PENDING/2) || (_pending == 0)) {
-                        _cond.notify_all();
-                    }
-                }
-                if ((lid % 256) == 0) {
-                    _attr.commit();
-                };
-            }));
-        }));
     }
-    void wait_complete() override {
-        std::unique_lock guard(_mutex);
-        while (_pending > 0) {
-            _cond.wait(guard);
+
+    void complete(uint32_t lid, std::unique_ptr<PrepareResult> prepared) {
+        _attr.setCommittedDocIdLimit(std::max(_attr.getCommittedDocIdLimit(), lid + 1));
```

Comment:
This assumes in-order completion. The documents with highest lid might not be visible after load due to a later
complete of a document with a lower lid when using out-of-order completions.